src/core/utils/hooks/useContainerWidth.tsx

```diff
@@ -0,0 +1,58 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import React from 'react';
+import { useMergedRefs } from './useMergedRefs';
+import { useResizeObserver } from './useResizeObserver';
+
+/**
+ * Hook that returns the width of an element in three stages:
+ *  - initialized with 0
+ *  - immediately set to element's initial width as soon as it's mounted
+ *  - update to new width every time it changes (using `useResizeObserver` hook)
+ *
+ * @private
+ * @param watchResizes If false, ResizeObserver will not be connected and only the initial width will be returned
+ * @returns [ref to attach to the element, stateful width of the element]
+ *
+ * @example
+ * const [ref, width] = useContainerWidth();
+ * ... // do something with width
+ * return <div ref={ref}>...</div>;
+ */
+export const useContainerWidth = <T extends HTMLElement>(
+  watchResizes = true,
+) => {
+  const [contentWidth, setContentWidth] = React.useState(0);
+
+  // callback ref to set initial contentWidth only once
+  const isInitialized = React.useRef(false);
+  const ref = React.useCallback((element: T) => {
+    if (!element) {
+      return;
+    }
+    if (!isInitialized.current) {
+      setContentWidth(element.getBoundingClientRect().width);
+    } else {
+      isInitialized.current = true;
+    }
+  }, []);
+
+  const updateWidth = React.useCallback(
+    ({ width }) => setContentWidth(width),
+    [],
+  );
+
+  const [resizeRef, resizeObserver] = useResizeObserver(updateWidth);
+
+  if (!watchResizes) {
```

Comment:
Didnt want to add this in useEffect?

Reply:
So it runs when watchResizes change only?

Reply:
This if condition is just to disconnect the stale resize observer

The real magic happens below where I'm not merging the resize ref:
![image](https://user-images.githubusercontent.com/9084735/142448587-679008f2-c0e6-4153-b74f-f800753b2eab.png)

Reply:
but if `watchResizes` is never changed, we expect just garbage collector to take care of disconnecting it?

Reply:
if `watchResizes` is never changed (meaning it's false) then `resizeObserver` would be undefined, so it wouldn't disconnect.

whereas if it changes from true to false, the manual disconnect is needed